src/map/script.cpp

```diff
@@ -19057,37 +19053,39 @@ BUILDIN_FUNC(getfreecell)
 	else
 		m = map_mapname2mapid(mapn);
 
-	map_search_freecell(NULL, m, &x, &y, rx, ry, flag);
+	if (sd)
+		map_search_freecell(&sd->bl, m, &x, &y, rx, ry, flag);
+	else
+		map_search_freecell(NULL, m, &x, &y, rx, ry, flag);
 
 	// Set MapX
-	num = st->stack->stack_data[st->start + 3].u.num;
-	name = get_str(num&0x00ffffff);
-	prefix = *name;
+	num_x = st->stack->stack_data[st->start + 3].u.num;
```

Comment:
Better would be to read the reference here already with `script_getdata(st,3);`
And then parse get the rest of the info with:
```
num = reference_getuid(data);
name = reference_getname(data);
```
And when setting the value at the end simply use data->ref.

This looks like someone who understands the core of the script engine wrote this, but noone else would...